### PR TITLE
Final group capstone - Book an Appointment

### PR DESCRIPTION
**Pull request by Gardimy**
Project Description: eSports Reservation System

**Overview**

This project involves developing an eSports reservation system with two distinct parts: the Backend, developed using Ruby on Rails, and the Frontend, built with React. This system aims to cater to eSports players, allowing them to reserve games, manage reservations, and access a catalog of available games.
 Click on the [link] (https://github.com/cancelei/EsportGamingCenter_frontend)  to navigate to the frontend of our APP.

Database Schema (Needs a revision, since the Normal user/Admin user is not clearly depicted in the document, and can generate doubts later):

- Esport Player - CurrentUser
- Esport Game - Can reserve a Game
- Gaming Center - Catalog of Available Games + Welcome Page
- Reservation Center - List of Reservations

Page Structure: 
Navigation (when the user is logged in): 
Games Catalog, 
Games List, (admin-only)
Reservations (list with delete button per item and button to Reserve(Form)), 
Reserve (form), 
Logout.
Navigation (when the user hasn't logged in): Login + Singup.

Core Features:

Shared tasks:

1 feat/Initialize Code in the Repository and Readme Preparation
 (Front and Backend)

Run Initialization commands in Front and backend.
Front: Run the npx create-react app or similar.
	Backend: Run rails new app-name -pg –api

Add Linters for Front and Backend
Add 2 Readme File that need to be connected with each other and there should be a “main” readme file where we will put the deployment link (We chose to be the Backend Repo).

2 feat/Responsive Design

The application should be responsive, with both mobile and desktop versions efficiently designed.

3 feat/Testing and Add Linters

Each person will test their respective tasks when this task is started.
Unit Test (Models, Controllers and Views)
Integration Test (For user interactions)

— Individual Tasks: —

1 feat/ [2pt] Individual Task - User Authentication and Login

Users log in via a simple interface, inputting only their username. The user table should have Full Name, email and let Devise gem create the remaining user table-related information.

Expected result after filling in the ‘Sign In’ form, the user should be redirected to the Catalog Page.

2 feat/ [2pt] Individual Task - Game Catalog

You will create a carousel that displays a Catalog of all Games available. (With name, short details, and genre).
Displays a carousel view of available games.

3 feat/ [2pt] Individual Task - Game Detail Page

Features a full description of the selected game.
Includes a 'Reserve' button to book the game.
This task will have code implementation on the controller, model, and view.
Use each game in the catalog as link to the correct Detail Page for that Game.

3.1 feat/ [2pt] Individual Task - Reserve from Detail Page

When a user goes to the ‘Game Detail Page’ and clicks on the Reserve Button, he should be prompted to the form where his email and the selected game are “hardcoded” on the form.

The expected result is that a new Reservation appear in the Reservations List.

3.1 feat/ [2pt] Individual Task - Reserve from Navigation.

When a user clicks the Reserve (in the nav bar) he should be prompted to the form where just his email is already selected (hardcoded).
Form fields: email, Game, Setup_config, and Date (MM/DD/YYYY). 

The user should be able to select all available games from the list.

The expected result is that a new Reservation appears in the Reservations List.

4 feat/ [2pt] Individual Task - Games List
Create a list of all available games. For now, just the basic structure is needed.

It should make all the games that are available in the Catalog appear here in List format, so in future tasks, we will transform it into an “Admin-related feature”.

4 feat/ [2pt] Individual Task - Adding New Game

A form is accessible via the navigation panel for adding new games to the catalog.

After adding an item with complete information needed according to the data base schema shared on 11/29/2023. The new Game should appear in the Catalog of Games.

This task will have code implementation on the controller, model and view.

An added game should appear in the Catalog Page and the Game List page.

5 feat/ [2pt] Individual Task - Deletion of Games

When navigating in the Games List with all games each row will have an add a  'Delete' button.

This task will have code implementation on the controller, model and view.

Expected Result: A deleted game should disappear from the Catalog Page and the Game List page.

7 feat/ [2pt] Individual Task - My Reservations Page

Users can view all their reservations, with details like game name, date, and config_PC.

8 feat/ [2pt] Individual Task - API Documentation

Comprehensive documentation for the API should be provided.

9 feat/ [2pt] Individual Task - Implement Restriction for Admin-Exclusive Access

Ensure that the ‘Game List’ view and methods  'Add Game' and 'Delete Game' are accessible only to users with admin privileges.

Validate if a User can delete his own reservations.
Validate if a User can create his own reservations.

Validate if the user and the admin have different permissions related to adding and deleting games. Important: A user shouldn’t be able to delete or add a new Game to the Catalog.

10 feat/ [2pt] Individual Task - Navigation Panel

You will develop a nav bar reviewing all of the already implemented routes, and create links to redirect users to the different views available.

Includes links to:
Game Catalog (carrousel)
Games List (list for Admins to add and delete)
Reserve (form)
Reservations (list) 
Sign Out

Optional Tasks:
11/  [2pt] Individual Task - API connection with External Esports data provider

5 Team members.

### CODE REVIEW

**Comment by cancelei** [comment]
@JuanDiegoGuerra we got any news from linter fixes?

File: config/database.yml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+  adapter: postgresql
+  encoding: unicode
+  username: <%= ENV['DB_USER'] %>
+  password: <%= ENV['DB_PSW'] %>
+  port: <%= ENV['DB_PORT'] %>
+  host: <%= ENV['DB_HOST'] %>
+
+development:
+  <<: *default
+  database: <%= ENV['DB_NAME_DEV'] %>
```
Comment (Meltrust) [inline config/database.yml:11]: - [x] This project relies on a .env file that has not been pushed to the repo:

image

Kindly, provide instructions about how to set up the required .env file with all the correct properties and their values.

That way, all your users will be able to setup and run your project correctly. 👍 

_I'll create my own in order to be able to continue the review._

File: README.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+./bin/dev
+```
+
+### Run tests
+
+To run tests, run the following command:
+
+```sh
+rspec ./spec/models
+```
```
Comment (Meltrust) [inline README.md]: - [x] The ./spec/models folder does not exist.  If you are not going to include models tests, kindly remove this section of the instructions.  👍

File: README.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+rails db:drop db:create db:migrate
+./bin/dev
+```
+
+### Run tests
+
+To run tests, run the following command:
+
+```sh
+rspec ./spec/models
```
Comment (Meltrust) [inline README.md]: - [x]  You are using rswag, but the "test" folder still exists, which is used by another test framework. Kindly consider deleting that folder if you are not using it. 👍

File: app/controllers/api/reservations_controller.rb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        render json: @reservation
+      else
+        render json: @reservation.errors, status: :unprocessable_entity
+      end
+    end
+
+    def destroy
+      @reservation.destroy
+      head :no_content
+    end
```
Comment (Meltrust) [inline app/controllers/api/reservations_controller.rb]: - [x] What happens here if the reservation is not successfully destroyed?  Kindly, add some error handling like you did in the update method (line 28). 👍

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Nice code organization ✔️ 
-  API is working well✔️ 
-  Good readme ✔️ 




#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

File: config/database.yml (hunk 1 above)
Comment (lerfast) [reply config/database.yml:11]: image
Done!!

**Review by Meltrust — APPROVED** [review]
# **Wow, you did it 🎉**
Brilliant
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
applause

### To highlight

- Superfluous test folder has been removed✔️ 
- Readme instructions have been vastly improved ✔️ 
- More consistent error handling has been implemented in controllers ✔️ 
- Superb job ✔️  



Cheers and Happy coding!👏👏👏


------

**Pull request closed without merging**